common/darkside.go

```diff
@@ -5,49 +5,176 @@ import (
+		// range, so we have to drop the existing range.
+		state.blocks = state.blocks[:0]
+	} else if blockHeight < state.startHeight {
+		// This block will replace the entire existing range.
+		state.blocks = state.blocks[:0]
+	} else {
+		// Drop the block that will be overwritten, and its children.
+		state.blocks = state.blocks[:blockHeight-state.startHeight]
+	}
+	if len(state.blocks) == 0 {
+		state.startHeight = blockHeight
+	} else {
+		// Set this block's prevhash.
+		prevblock := parser.NewBlock()
+		rest, err := prevblock.ParseFromSlice(state.blocks[len(state.blocks)-1])
 		if err != nil {
-			Log.Fatal("Error loading default darksidewalletd blocks")
+			return err
 		}
-		scan := bufio.NewScanner(testBlocks)
-		for scan.Scan() { // each line (block)
-			block := scan.Bytes()
-			state.blocks = append(state.blocks, string(block))
+		if len(rest) != 0 {
+			return errors.New("block is too long")
 		}
+		copy(blockData[4:4+32], prevblock.GetEncodableHash())
+	}
+	if state.saplingActivation < 0 {
+		state.saplingActivation = blockHeight
 	}
+	state.blocks = append(state.blocks, blockData)
+	return nil
+}
+
+func readBlocks(src io.Reader) error {
+	// some blocks are too large, especially when encoded in hex, for the
+	// default buffer size, so set up a larger one; 8mb should be enough.
+	scan := bufio.NewScanner(src)
+	var scanbuf []byte
+	scan.Buffer(scanbuf, 8*1000*1000)
+	for scan.Scan() { // each line (block)
+		if err := DarksideAddBlock(scan.Text()); err != nil {
+			return err
+		}
+	}
+	if scan.Err() != nil {
+		return scan.Err()
+	}
+	return nil
+}
 
-	if time.Now().Sub(state.server_start).Minutes() >= 30 {
+func DarksideSetMetaState(sa int32, bi, cn string) error {
+	state.saplingActivation = int(sa)
+	state.branchID = bi
+	state.chainName = cn
+	return nil
+}
+
+func DarksideGetIncomingTransactions() [][]byte {
+	return state.incomingTransactions
+}
+
+func DarksideSetBlocksURL(url string) error {
+	if strings.HasPrefix(url, "file:") && len(url) >= 6 && url[5] != '/' {
+		dir, err := os.Getwd()
+		if err != nil {
+			return err
+		}
+		url = "file:" + dir + string(os.PathSeparator) + url[5:]
+	}
+	cmd := exec.Command("curl", "--silent", "--show-error", url)
+	stdout, err := cmd.StdoutPipe()
+	if err != nil {
+		return err
+	}
+	stderr, err := cmd.StderrPipe()
+	if err != nil {
+		return err
+	}
+	if err := cmd.Start(); err != nil {
+		return err
+	}
+	defer cmd.Wait()
+	err = readBlocks(stdout)
+	if err != nil {
+		return err
+	}
+	stderroutstr, err := ioutil.ReadAll(stderr)
+	if err != nil {
+		return err
+	}
+	if len(stderroutstr) > 0 {
+		return errors.New(string(stderroutstr))
+	}
+	return nil
+}
+
+func DarksideSendTransaction(txbytes []byte) {
+	state.incomingTransactions = append(state.incomingTransactions, txbytes)
+}
+func darksideRawRequest(method string, params []json.RawMessage) (json.RawMessage, error) {
+	if time.Now().Sub(state.serverStart).Minutes() >= 30 {
```

Comment:
At at glance at the code, it isn't obvious to me that this check happens for *every* darkside operation, e.g. I think `DarksideSetMetaState` or `DarksideSetBlocksURL` could be executed without triggering the check. Let's extract this into a helper method called `darksideShutdownIfRunningTooLong` (or similar) and call it at the start of all the public functions in this file so it's really obvious the check always happens.

Reply:
Nice!